### PR TITLE
Support dockerized builds (from source)

### PR DESCRIPTION
**Pull request by amcgee**
This is a simple non-breaking change (adding a dockerfile) which adds support for zero-dependency builds from source (or single-dependency if you include the Docker platform).  There are numerous benefits to this approach, a few outlined below:

1. Dependencies are fully encapsulated, codified, and deterministic.  To update the required version of maven, for instance, or a Java dependency there's not risk that developers will need to notice and act on that upgrade in their local environment.
2. The build environment is fully isolated from machine and platform context - it doesn't matter what version of Java, Maven, Git, node/npm, or Tomcat is installed locally (none are required) to build a `dhis2.war` from source and serve it from a dockerized Tomcat server.
3. The built docker image is portable and self-contained, which allows it to easily be deployed in develpment, testing, and customer environments.  There are other examples out there but see [this docker-compose setup](github.com/amcgee/dhis2-backend) for an example.
4. Once this file is in place we can push built Docker images to hub.docker.com.  Then installation of a fully-functional DHIS2 instance could be as simple as `docker up --rm dhis2/core:2.31.0-alpine` **Note**: there are already two docker hub DHIS2 image repositories - [dhis2/dhis2-web](hub.docker.com/r/dhis2/dhis2-web) (community-run, possibly out of date, and packaged from the built WAR file rather than built from source) and [amcgee/dhis2-core](hub.docker.com/r/amcgee/dhis2-core) (built from source in my fork of dhis2-core in developing this Dockerfile, should go away when this is merged).  We will want to consolidate these under the `dhis2` org on [hub.docker.com](hub.docker.com).
5. Co-locating the Dockerfile with code ensures that the Docker build process is synchronized with the  latest core source (and can also be branched to maintain synchronized with different release versions).  It also follows Docker best-practices which strongly recommend building images from within the Dockerfile build context.

---

Some documentation (an abbreviated version of which is included as comments in the Dockerfile and in CONTRIBUTING.md

# Building and Running DHIS2_Core with Docker

DHIS2 Core can be built as a Docker image and run entirely withing a Docker Container.  This approach is an alternative to building and running with local versions of Maven, JVM, and Tomcat.

## Install Docker

There is only one pre-requisite to build and run from source:

- [Docker Desktop](https://www.docker.com/products/docker-desktop) (OSX and Windows)
- OR [Docker Engine](https://docs.docker.com/install/) (Linux and friends)

To learn the basic principles of Docker, see the official [Getting Started](https://docs.docker.com/get-started/) and [Overview](https://docs.docker.com/engine/docker-overview/) guides

## Build a Docker image

You can specify anything you'd like as the `<imagetag>` or omit it.  We will use `dhis2-core:local-dev-alpine` here, which has the following elements:

- `dhis2-core` as a made-up image repository
- `local` since we're building a custom image
- `dev` for the release we're building
- `alpine` to indicate the flavor of linux OS upon which we've based our image

This image tag identifier format follows the format used by [Docker Hub](https://hub.docker.com) which hosts pre-built images.  When you build and tag an image locally it is (by default) only cached on your local machine, so you can name it whatever you'd like.  See 

```sh
# Generically
> docker build [-t <imagetag>] .

# For example
> docker build -t dhis2-core:local-dev-alpine .

# Or just build without specifying a tag
> docker build .
```

Docker will construct a fresh Linux container and install all build dependencies, Maven will build `dhis.war`, and finally Docker will build a slim runtime container with only Tomcat and the built application.  See the `Dockerfile` for details of what this does under the hood.

## Run DHIS2-Core in Docker

Once you have built your image, you can tell Docker to spin up a DHIS2-Core container and run it as a daemon.

You can specify whatever you'd like for the variables `<port>`, `<configdir>`, and `<image>`.

- `<port>` is the port on your host machine to bind to the running DHIS2-Core service (usually 8080)
- `<configdir>` is the path to a DHIS2_home directory on your host machine.  `:ro` is specified to mount the directory as read-only
- `<image>` will likely be the `<imagetag>` tag you specified above, but it can also point to a pre-built DHIS2-core image hosted on [Docker Hub](https://hub.docker.com), such as [amcgee/dhis2-core:2.30-alpine](https://hub.docker.com/u/amcgee/dhis2-core)

```sh
# Generically
> docker run -d --rm -p <port>:80 -v <configdir>:/DHIS2_home:ro <image>

# For example (defaulting to ./dhis2 if DHIS2_home is set in the environment)
> docker run -d --rm -p 8080:80 -v ${DHIS2_home:-./dhis2}:/DHIS2_home:ro dhis2-core:local-dev-alpine

# Or specify a tagged 2.30 release image on Docker Hub - Docker will pull the base image automatically
> docker run -d --rm -p 8080:80 -v ./dhis2_home:/DHIS2_home:ro amcgee/dhis2-core:2.30-alpine
```

This will output a `<containerid>`, you will need this ID to manage the running service.

> **Note** You can find the `<containerid>` later by running `docker ps`

The DHIS2 Core instance should now be running in the background and accessible at [localhost:8080](http://localhost:8080) (or on whatever `<port>` you specified)

## Manage the running Docker Container

You can tail the logs of the running container:

```sh
> docker logs --follow <containerid>
```

Or kill it:

```sh
> docker kill <containerid>
```

### CODE REVIEW

**Comment by vilkg** [comment]
@jason - I don't think we should bake in postgres to core image, i would say it should be separate image. Different devs might have different database setup, they should be free to use any db instance they want.

**Comment by jason-p-pickering** [comment]
@vilkg Yes, I get it now. This solves the problem of actually being able to build the artifact. Got it. Looks good to and think it will be very useful @amcgee

**Comment by vilkg** [comment]
@amcgee  - I haven't tested it yet, but I´m going to try to do it today. Did you intend it to be for local use only? From CI point of view, I would imagine the following pipeline: 
- build war (mvn clean install)
- Run integration and api tests
- Push image to docker hub. 

For this to be efficient, maybe we shouldn't build war and run all the tests again while building an image?

**Comment by amcgee** [comment]
> @amcgee - I haven't tested it yet, but I´m going to try to do it today. Did you intend it to be for local use only? From CI point of view, I would imagine the following pipeline:
> 
> * build war (mvn clean install)
> * Run integration and api tests
> * Push image to docker hub.
> 
> For this to be efficient, maybe we shouldn't build war and run all the tests again while building an image?

@vilkg I originally had the build skip tests, but I think it's actually important to run them if we're publishing the artifact we create.

I might actually argue in favor of **only** building in Docker.  It encapsulates our dependencies and is effectively a duplicate of everything we're doing outside of Docker today.  The end result should be the same (dhis.war) but built in the containerized environment.  The advantage there is also that the CI build becomes 100% repeatable in any environment, we have a docker image available to spin up and run the api and integration tests, and we're not duplicating any maven command specifications (currently duplicated in `.travis.yml`) or build effort since the built WAR should be 100% portable.

I think it's important to build the Docker Image from source, rather than packaging a pre-built WAR, which is why I did it this way.  It's certainly possibly to just bundle the WAR with tomcat but that doesn't do a whole ton for us.

So the CI process I propose would be would be:

* build (`docker build`, including unit tests)
* test (api and integration)
** `docker run` (spin up the server, pointing at a database which we might need to seed)
** Run api/integration tests
* Publish artifacts
** Docker image to docker hub
** WAR file to s3

A few potential downsides of this approach:

* We *may* want to publish different base-image versions of DHIS2 docker image, which would involve splitting up the Dockerfile into Build and Run variants, rather than a single file with two stages
* I'm not sure what the concurrency model is in Travis with `sudo: true` which would be needed to build and run containers.  This could mean performance slows down compared to the current Travis builds, but I imagine they can be tuned to perform similarly since normal Travis builds run on Docker anyway.

A final consideration would be Jenkins vs Travis - in my limited understanding of the Core build pipeline both Jenkins and Travis are running.  Should we build a docker image in one or the other or both?  Ideally the build process should be identical between Travis and Jenkins, I don't know if it is today.

**Comment by amcgee** [comment]
> @amcgee I think this definitely serves a purpose to allow for one to build the artifact in a stable environment. However, it seems to be missing the Postgresql backend or maybe I missed it?

Yep, as @vilkg said the database should live separately so that each container has a singular purpose.  They can then be composed together into a cluster of microservices using something simple like `docker-compose` or something more robust and scalable like `docker swarm` or `kubernetes` (which we don't officially support but which could conceivably be built using this docker image as a primitive).

The sample `docker-compose` I use is [here](https://github.com/amcgee/dhis2-backend) and includes 3 images - DHIS2, NGINX, and Postgres - with appropriate volume and port bindings to make them easy to run in a local dev environment.

**Comment by vilkg** [comment]
@amcgee - that pipeline sounds good.

Regarding Jenkins vs travis - travis pipeline is running as a check for PR, but we mainly use jenkins. Jenkins runs tests, triggers SQ analysis, builds wars.

**Comment by Philip-Larsen-Donnelly** [comment]
I'm afraid I've been off sick for a few days and out of the loop, but this is really interesting @amcgee! Good stuff!
In principle I see no problem in merging this PR as it simply adds some instructions and configuration to allow optional use of docker in a simple way, but I will leave the decision to the "masters" of core!
I look forward to checking this out in detail in early January and discussing potential CI and delivery pipelines!

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+##########
+# BUILD STAGE 2
+# Serve the packaged .war file (Tomcat)
+# Use a second build stage in a fresh container so we don't bloat it with the Maven build tools
+##########
+FROM tomcat:8.5.34-jre8-alpine as serve
+
+RUN rm -rf /usr/local/tomcat/webapps/*
+COPY --from=build /src/dhis-web/dhis-web-portal/target/dhis.war /usr/local/tomcat/webapps/ROOT.war
```
[the remark this answers is not in the document]
Comment (amcgee) [reply Dockerfile]: @vilkg Sorry about that!  I updated the ADD in the build stage but missed the SERVE stage reference.  Fixed locally, building to confirm.

**Comment by Philip-Larsen-Donnelly** [comment]
Merged. It is zero risk and is easy to remove if desired.

**Pull request merged**